### PR TITLE
test(st): right-size too-tight golden tolerances across s-tests (bf16/fp16/fp32-K-split)

### PR DESCRIPTION
**Pull request by tonibohnlein**
## Summary
Several device s-tests validate **reduced-precision** compute against a golden with a tolerance **tighter than one ULP of the precision the kernel actually computes in** — most at the default `rtol=atol=1e-5`, some at a bf16-insufficient `1e-3`. That guarantees intermittent single-element flakes whenever a value lands on a rounding boundary (the device's hardware `exp`/`sqrt`/matmul-reduction rounds differently than torch). This PR right-sizes each to admit ≥ 1 ULP of its precision.

This started from `batch_softmax_prepare` (bf16 softmax at `1e-5`) and, after an audit of `tests/st/runtime/`, was extended to the rest of the same flake class — two of which have since flaked live in CI (`tensor_batch_matmul` fp32 K-split; `tile_sqrt_fp16`).

## Tolerance policy (admit ≥ 1 ULP of the compute precision)
| class | rtol / atol | rationale |
|---|---|---|
| **bf16** arithmetic / output | `1.6e-2 / 1e-3` (softmax) · `2e-2 / 1e-3` (attention) | bf16 `ε = 2⁻⁷ ≈ 7.8e-3`; one ULP up to `2⁻⁸ ≈ 3.9e-3`. `1.6e-2` = PyTorch bf16 default; `2e-2` matches existing bf16 attention s-tests |
| **fp16** transcendental | `2e-3 / 2e-3` | fp16 `sqrt` HW Newton approx differs from `torch.sqrt` by ~1–2 fp16 ULP (max ULP near 1 is `2⁻¹⁰ ≈ 9.8e-4`). `2e-3` clears the observed flake (`2⁻¹¹ ≈ 4.9e-4`) with ~8× margin — `1e-3` gave only ~1 ULP, too thin |
| **fp32 K-split matmul** | `1e-4 / 1e-4` | device K-reduction order ≠ single-pass torch golden → `~K·eps_fp32 ≈ 1.5e-5` at K=128 (larger for cancellation-heavy elements). Same class the AutoL0 matmul s-tests already loosen |

## Tests fixed
- **bf16** — `batch_softmax_prepare` (1.6e-2); `test_paged_attention_multi_config::SoftmaxPrepare` (+ its PTOAS subclass, 1.6e-2); `test_batch_paged_attention::BatchPagedAttention` full attention (2e-2 — BF16 QK/PV matmuls dominate; its softmax `pij` intermediate is FP16); `test_qwen3_decode_scope3_mixed` (2e-2). Tolerance sets guarded by `if kwargs.get("config") is None` so a caller-supplied config is respected.
- **fp16** — `test_unary_math::TileSqrtTestCase` fp16 sqrt (set in `_UnaryMathBase.__init__` for `dtype==FP16`, so it also covers any future fp16 sin/cos).
- **fp32 K-split** — `test_mat_slice_to_left`; all 5 `test_tensor_batch_matmul` fp32 batch-matmul cases.

Each is a one-line tolerance edit (`self.config.rtol/atol` or the `RunConfig(...)` literal). **No kernel or codegen change.**

## Deliberately not changed (audit found these safe)
Pure **data-movement** bf16/fp16 outputs (scatter/gather/copy/pad/reshape/transpose) are device-exact; **device-order-matched** goldens (e.g. `test_col_reduction` replicates the fp16 reduction tree); **constant/integer-input** matmuls (bf16-exact products); and ops already appropriately loosened. Two fp16 elementwise cases (`test_vector_misc` muls / col_expand_add) are borderline — a single fp16 mul/add is normally bit-exact — so they're left as-is pending an actual flake, to avoid loosening genuinely-exact ops.

## Why the earlier reviewed part changed
The `batch_softmax_prepare` value was corrected from `1e-3` up to `1.6e-2` after review (a `1e-3` threshold at 0.5 is `1.5e-3`, below one bf16 ULP of `3.9e-3`). The same reasoning is applied consistently across the bf16 tests here.

### CODE REVIEW

File: tests/st/runtime/framework_and_models/test_batch_paged_attention.py [hunk 1]
```diff
@@ -233,6 +233,18 @@ def __init__(
...
+        # The device's hardware exp is a ~1-2 ULP approximation, so an exp value
+        # near a BF16 rounding boundary lands on a different BF16 than the
+        # torch.exp golden, shifting one summand by one BF16 ULP (2^-10 ~ 9.8e-4).
+        # BF16 has 8 significand bits (u = 2^-8 ~ 3.9e-3), so the default 1e-5 is
+        # ~2 orders of magnitude tighter than a single BF16 ULP and flakes
+        # intermittently. Matches the 1e-3 already used by the full-attention
+        # integration case in this file (and still tighter than PyTorch's 1.6e-2
+        # bf16 default).
+        self.config.atol = 1e-3
+        self.config.rtol = 1e-3
```
[the remark this answers is not in the document]
Comment (tonibohnlein) [reply tests/st/runtime/framework_and_models/test_batch_paged_attention.py]: Good catch — addressed in `0d711ec9`.

You're right that `rtol=atol=1e-3` doesn't cover a full BF16 ULP for values near 0.5: a summand/`pij_batch` element in `[0.5, 1)` has ULP `2^-8 ≈ 3.9e-3`, while `1e-3` only admits `1e-3 + 1e-3·0.5 = 1.5e-3` there — so the hardware-exp-vs-torch-exp flake would survive for different random inputs.

Fix: `rtol` now covers ≥ one BF16 ULP *relative* everywhere. BF16 `eps = 2^-7 ≈ 7.8e-3` (max relative ULP, at the low end of each binade), so I set **`rtol=1.6e-2`** (PyTorch's bf16 `assert_close` default, ~2-ULP margin) with **`atol=1e-3`** for near-zero summands. At the 0.5 worst case the threshold is now `1e-3 + 1.6e-2·0.5 = 9e-3 > 3.9e-3` (covers even 2 ULP). The comment and commit message were updated to reflect the ULP analysis.

**Review by Hzfengsy — APPROVED** [review]
Test tolerance right-sizing looks sound; the BF16-ULP tolerance concern from the automated reviewer was addressed. LGTM.
